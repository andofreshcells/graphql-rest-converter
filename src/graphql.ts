import _ from 'lodash'
import {
  DocumentNode,
  GraphQLSchema,
  visit,
  print,
  Kind,
  ASTNode,
  buildASTSchema,
  OperationDefinitionNode,
  GraphQLDirective,
  OperationTypeNode,
  validate,
} from 'graphql'
import { mergeSchemas } from '@graphql-tools/schema'
import { gql } from 'graphql-tag'
import { OpenAPIV3 } from 'openapi-types'
import { getVariablesFromPathTemplate } from './pathTemplate'
import {
  createFragmentMap,
  getDependencyClosure,
  getDirective,
  getDirectiveArguments,
  getFragmentDependencies,
  getReferencedFragments,
  isOperationDefinitionNode,
} from './graphqlUtils'
import { CustomProperties, OAType, BridgeOperation } from './types'
import { getReferenceableFragments, GraphQLTypeToOpenAPITypeSchemaConverter } from './typeConverter'
import { isNullable } from './openApi'
import { printSourceLocation } from 'graphql/language/printLocation'

const DIRECTIVE_DEFINITION = gql`
  input OAExternalDocsInput {
    url: String!
    description: String
  }

  # see https://swagger.io/docs/specification/authentication/
  input SecurityDefinitionInput {
    schema: String!
    scopes: [String!]
  }

  enum HttpMethod {
    GET
    POST
    PUT
    DELETE
  }

  enum ParameterSource {
    PATH
    QUERY
    HEADER
  }

  directive @OAOperation(
    path: String!
    tags: [String!]
    summary: String
    security: [SecurityDefinitionInput]
    description: String
    externalDocs: OAExternalDocsInput
    deprecated: Boolean
    "defaults to GET"
    method: HttpMethod
  ) on QUERY | MUTATION

  directive @OAParam(
    in: ParameterSource
    deprecated: Boolean
    description: String
    required: Boolean
    name: String
  ) on VARIABLE_DEFINITION

  directive @OABody(description: String) on VARIABLE_DEFINITION

  directive @OADescription(
    description: String
  ) on FRAGMENT_DEFINITION | FIELD | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
`

export enum OpenAPIDirectives {
  Operation = 'OAOperation',
  Param = 'OAParam',
  Body = 'OABody',
  Description = 'OADescription',
}

const graphqlOperationDirectiveDataToOpenAPIOperation = (
  requestConfig: OpenAPIOperationDirectiveData
) => {
  const { tags, summary, description, externalDocs, security, deprecated } = requestConfig
  return {
    ...(tags?.length ? { tags } : {}),
    ...(typeof summary === 'string' ? { summary } : {}),
    ...(typeof description === 'string' ? { description } : {}),
    ...(typeof externalDocs?.url === 'string' ? { externalDocs } : {}),
    ...(deprecated ? { deprecated } : {}),
    ...(security
      ? {
          security: security.map((item) => {
            if (item === null) {
              return {}
            }
            return { [item.schema]: item.scopes || [] }
          }),
        }
      : {}),
  }
}

const getOpenAPIParameters = (
  variablesSchema: Record<string, OAType>,
  path: string,
  paramsDirectiveData: Record<string, OpenAPIParamDirectiveData>
) => {
  const parameters: OpenAPIV3.ParameterObject[] = []
  const variableMap: Record<string, string> = {}

  const pathVariables = new Set(getVariablesFromPathTemplate(path))

  for (const [variableName, paramDirectiveData] of Object.entries(paramsDirectiveData)) {
    let parameterName = variableName
    const nameOverride = paramDirectiveData.name
    if (nameOverride) {
      parameterName = nameOverride
    }
    const paramDirectiveDataIn = paramDirectiveData.in?.toLowerCase()

    if (paramDirectiveDataIn) {
      if (pathVariables.has(parameterName) && paramDirectiveDataIn !== 'path') {
        throw new Error(
          `Location ${paramDirectiveDataIn} invalid for parameter ${parameterName} because it is part of the path`
        )
      }
      if (!pathVariables.has(parameterName) && paramDirectiveDataIn === 'path') {
        throw new Error(
          `Location ${paramDirectiveDataIn} invalid for parameter ${parameterName} because it is not part of the path`
        )
      }
      if (paramDirectiveDataIn === 'cookie') {
        throw new Error(`Unsupported parameter location cookie for parameter ${parameterName}`)
      }
      if (!['path', 'query', 'header'].includes(paramDirectiveDataIn)) {
        throw new Error(
          `Unknown parameter location ${paramDirectiveDataIn} for parameter ${parameterName}`
        )
      }
    }
    if (paramDirectiveDataIn && paramDirectiveDataIn === 'header') {
      parameterName = parameterName.toLowerCase()
    }
    if (parameterName !== variableName) {
      variableMap[parameterName] = variableName
    }

    const in_ = pathVariables.has(parameterName)
      ? 'path'
      : paramDirectiveDataIn && ['header', 'cookie'].includes(paramDirectiveDataIn)
      ? paramDirectiveDataIn
      : 'query'
    const deprecated = paramDirectiveData.deprecated
    const description = paramDirectiveData.description
    const required = paramDirectiveData.required
    const schema = variablesSchema[variableName]
    parameters.push({
      in: in_,
      name: parameterName,
      schema,
      ...(isNullable(schema) ? { required: true } : {}),
      ...(nameOverride ? { [CustomProperties.VariableName]: variableName } : {}),
      ...(deprecated === true ? { deprecated } : {}),
      ...(description ? { description } : {}),
      // path parameters are always required (see https://swagger.io/docs/specification/describing-parameters/)
      ...(in_ === 'path' ? { required: true } : { required: required || false }),
    })
  }

  return {
    parameters,
    variableMap,
  }
}

const getOpenAPIRequestBody = (
  variablesSchema: Record<string, OAType>,
  variableName: string,
  bodyDirectiveData: OpenAPIBodyDirectiveData
) => {
  const description = bodyDirectiveData.description
  return {
    content: {
      'application/json': {
        schema: variablesSchema[variableName],
      },
    },
    ...(description ? { description } : {}),
    [CustomProperties.VariableName]: variableName,
  }
}

type OpenAPIOperationDirectiveData = {
  path: string
  tags?: [string]
  summary?: string
  description?: string
  externalDocs?: {
    url: string
    description?: string
  }
  security?: [
    {
      schema: string
      scopes: string[]
    } | null
  ]
  deprecated?: boolean
  method: OpenAPIV3.HttpMethods
}

type OpenAPIParamDirectiveData = {
  in?: string
  deprecated?: boolean
  description?: string
  required?: boolean
  name?: string
}

type OpenAPIBodyDirectiveData = {
  description?: string
}

const createOpenAPIOperation = (
  operationId: string | null,
  operationSource: string,
  parameters: OpenAPIV3.ParameterObject[],
  requestBody: OpenAPIV3.RequestBodyObject | null,
  resultSchema: OAType,
  operationDirectiveData: OpenAPIOperationDirectiveData
) => {
  const responses = {
    '200': {
      description: 'Success',
      content: {
        'application/json': {
          schema: resultSchema,
        },
      },
    },
    '400': {
      description: 'Invalid request',
      content: {
        'application/json': {
          schema: {
            type: 'object',
            properties: {
              errors: {
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    errorCode: { type: 'string' },
                    location: { type: 'string' },
                    message: { type: 'string' },
                    path: { type: 'string' },
                  },
                },
              },
            },
          } as OpenAPIV3.SchemaObject,
        },
      },
    },
    '500': {
      description: 'Internal Server Error',
      content: {
        'application/json': {
          schema: {
            type: 'object',
            properties: {
              data: {
                ...resultSchema,
                nullable: true,
                description: 'Branch of data that does not contain errors',
              },
              errors: {
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    locations: {
                      type: 'array',
                      items: {
                        type: 'object',
                        properties: {
                          line: { type: 'integer' },
                          column: { type: 'integer' },
                        },
                      },
                    },
                    message: { type: 'string' },
                    path: { type: 'array', items: { type: 'string' } },
                  },
                },
              },
            },
          } as OpenAPIV3.SchemaObject,
        },
      },
    },
  }
  return {
    ...(operationId !== null ? { operationId } : {}),
    [CustomProperties.Operation]: operationSource,
    ...graphqlOperationDirectiveDataToOpenAPIOperation(operationDirectiveData),
    parameters,
    ...(requestBody !== null ? { requestBody } : {}),
    responses,
  }
}

const getVariablesDirectiveData = (
  paramDirectiveDefinition: GraphQLDirective,
  bodyDirectiveDefinition: GraphQLDirective,
  operation: OperationDefinitionNode
) => {
  const paramsDirectiveData: Record<string, OpenAPIParamDirectiveData> = {}
  const bodiesDirectiveData: Record<string, OpenAPIParamDirectiveData> = {}

  for (const variableDefinition of operation.variableDefinitions || []) {
    const bodyDirective = getDirective(variableDefinition, OpenAPIDirectives.Body)
    if (bodyDirective) {
      bodiesDirectiveData[variableDefinition.variable.name.value] = getDirectiveArguments(
        bodyDirectiveDefinition,
        bodyDirective
      ) as OpenAPIBodyDirectiveData
      continue
    }
    const paramDirective = getDirective(variableDefinition, OpenAPIDirectives.Param)
    paramsDirectiveData[variableDefinition.variable.name.value] = paramDirective
      ? (getDirectiveArguments(
          paramDirectiveDefinition,
          paramDirective
        ) as OpenAPIParamDirectiveData)
      : {}
  }

  return {
    paramsDirectiveData,
    bodiesDirectiveData,
  }
}

const directiveSchema = buildASTSchema(DIRECTIVE_DEFINITION)

export const getBridgeOperations = (
  schema: GraphQLSchema,
  document: DocumentNode,
  customScalars?: (scalarTypeName: string) => OpenAPIV3.SchemaObject
) => {
  const bridgeOperations: Array<BridgeOperation> = []

  const fragmentMap = createFragmentMap(document.definitions)

  const fragmentDependencies = getFragmentDependencies(fragmentMap)

  const referencableFragments = getReferenceableFragments(schema, fragmentMap, document)

  const typeConverter = new GraphQLTypeToOpenAPITypeSchemaConverter(
    schema,
    customScalars,
    fragmentMap,
    referencableFragments
  )
  const validationResult = validate(mergeSchemas({ schemas: [directiveSchema, schema] }), document)
  if (validationResult.length > 0) {
    const errors = validationResult
      .map(
        (error) =>
          `${error.message} Source: ${
            error.locations && error.source
              ? printSourceLocation(error.source, error.locations[0])
              : 'unknown'
          }`
      )
      .join(', ')
    throw new Error(`Schema validation error(s): ${errors}`)
  }

  for (const operation of document.definitions) {
    if (!isOperationDefinitionNode(operation)) {
      continue
    }

    const operationDirective = getDirective(operation, OpenAPIDirectives.Operation)
    if (!operationDirective) {
      continue
    }

    // # Extract all directives

    const operationDirectiveData = getDirectiveArguments(
      directiveSchema.getDirective(OpenAPIDirectives.Operation)!,
      operationDirective
    ) as OpenAPIOperationDirectiveData

    const { paramsDirectiveData, bodiesDirectiveData } = getVariablesDirectiveData(
      directiveSchema.getDirective(OpenAPIDirectives.Param)!,
      directiveSchema.getDirective(OpenAPIDirectives.Body)!,
      operation
    )

    // # Compile stand-alone document

    // ## Remove custom directives

    const operation_ = operation

    // ## Include dependency fragments

    const operationFragmentDependencies = getDependencyClosure(
      getReferencedFragments(operation_),
      fragmentDependencies
    )
    const referencedFragments = Object.values(
      _.pickBy(fragmentMap, (v, k) => operationFragmentDependencies.has(k))
    )

    const singleOperationDocument = {
      kind: Kind.DOCUMENT,
      definitions: [operation_, ...referencedFragments],
    } as DocumentNode

    const operationId = operation_.name ? operation_.name.value : null

    // # Extract types from GraphQL operation

    const variablesSchema = typeConverter.variablesFromOperation(operation_)

    const resultSchema = typeConverter.resultFromOperation(operation_)

    // # Build OpenAPI schema

    // ## Build OpenAPI schema: Parameters

    const { parameters, variableMap } = getOpenAPIParameters(
      variablesSchema,
      operationDirectiveData.path,
      paramsDirectiveData
    )

    // ## Build OpenAPI schema: Request body

    const bodyVariables = Object.keys(bodiesDirectiveData)
    if (bodyVariables.length > 1) {
      throw new Error('Only one "OABody" variable allowed')
    }
    const requestBodyVariable = bodyVariables.length === 1 ? bodyVariables[0] : null
    const requestBody = requestBodyVariable
      ? getOpenAPIRequestBody(
          variablesSchema,
          requestBodyVariable,
          bodiesDirectiveData[requestBodyVariable]
        )
      : null

    const graphqlDocument = removeOpenAPIDirectives(singleOperationDocument)
    const operationSource = print(graphqlDocument)

    // ## Build OpenAPI schema: Operation

    const openAPIOperation = createOpenAPIOperation(
      operationId,
      operationSource,
      parameters,
      requestBody,
      resultSchema,
      operationDirectiveData
    )

    // # Build bridge operation

    const defaultHttpMethod =
      operation_.operation === OperationTypeNode.MUTATION
        ? OpenAPIV3.HttpMethods.POST
        : OpenAPIV3.HttpMethods.GET

    const httpMethod =
      (operationDirectiveData.method?.toLowerCase() as OpenAPIV3.HttpMethods | undefined) ??
      defaultHttpMethod

    const bridgeOperation = {
      openAPIOperation,
      path: operationDirectiveData.path,
      httpMethod,
      graphqlDocument,
      graphqlDocumentSource: operationSource,
      variableMap,
      requestBodyVariable,
    }

    bridgeOperations.push(bridgeOperation)
  }

  return {
    operations: bridgeOperations,
    schemaComponents: typeConverter.getSchemaComponents(),
  }
}

export const removeOpenAPIDirectives = <T extends ASTNode>(node: T): T => {
  return visit(node, {
    Directive: {
      enter(node) {
        if (Object.values(OpenAPIDirectives).includes(node.name.value as OpenAPIDirectives)) {
          return null
        }
      },
    },
  })
}
