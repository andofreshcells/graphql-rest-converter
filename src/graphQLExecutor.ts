import { execute, parse, GraphQLSchema, GraphQLError, ExecutionResult } from 'graphql'
import { ClientError, GraphQLClient } from 'graphql-request'

export type GraphQLExecutorArgs = {
  document: string
  variables: { [key: string]: any }
}

export type GraphQLExecutor = (args: GraphQLExecutorArgs) => Promise<ExecutionResult>

export const createHttpExecutor = (
  ...args: ConstructorParameters<typeof GraphQLClient>
): GraphQLExecutor => {
  const client = new GraphQLClient(...args)
  return async ({ document, variables }) => {
    try {
      // TODO: Use `errorPolicy: 'all'` when released
      return client.rawRequest({ query: document, variables })
    } catch (e) {
      if (e instanceof ClientError) {
        return e.response as ExecutionResult
      }
      throw e
    }
  }
}

export const createSchemaExecutor = (schema: GraphQLSchema): GraphQLExecutor => {
  return ({ document, variables }) =>
    Promise.resolve(execute({ schema, document: parse(document), variableValues: variables }))
}
