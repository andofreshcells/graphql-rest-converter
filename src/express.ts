/* eslint-disable @typescript-eslint/ban-ts-comment */
import _ from 'lodash'
import express, { RequestHandler, NextFunction, IRouter, Request, Response } from 'express'
import bodyParser from 'body-parser'
import { OpenAPIV3 } from 'openapi-types'
import { parse, buildSchema, print } from 'graphql'
import OpenAPIRequestCoercer from 'openapi-request-coercer'
import OpenAPIRequestValidator from 'openapi-request-validator'
import OpenAPIResponseValidator from 'openapi-response-validator'
import { getBridgeOperations } from './graphql.js'
import { pathTemplateToExpressRoute } from './pathTemplate.js'
import {
  BridgeOperation,
  BridgeOperations,
  SchemaComponents,
  CustomOperationProps,
  CreateOpenAPIGraphQLBridgeConfig,
  CreateOpenAPISchemaConfig,
  CreateMiddlewareConfig,
} from './types.js'
import { GraphQLExecutor } from './graphQLExecutor.js'
import RequestBodyObject = OpenAPIV3.RequestBodyObject
import { InvalidResponseError } from './errors.js'
import { createOpenAPISchemaWithValidate, resolveSchemaComponents } from './utils.js'
import { transformRequest } from './multipart.js'

const middlewareToPromise =
  (middleware: RequestHandler) =>
  (req: Parameters<RequestHandler>[0], res: Parameters<RequestHandler>[1]) => {
    return new Promise<ReturnType<RequestHandler>>((resolve, reject) => {
      const next = (x: unknown) => (x ? reject(x) : resolve())
      middleware(req, res, next)
    })
  }

const promiseToHandler =
  (
    handler: (
      req: Parameters<RequestHandler>[0],
      res: Parameters<RequestHandler>[1]
    ) => Promise<ReturnType<RequestHandler>>
  ) =>
  (...args: Parameters<RequestHandler>) => {
    const result = handler(args[0], args[1])
    result.catch(args[2] as NextFunction)
  }

const jsonBodyParserPromise = middlewareToPromise(bodyParser.json())
const formBodyParserPromise = middlewareToPromise(bodyParser.urlencoded())

const addOperation = <
  T extends CustomOperationProps = CustomOperationProps,
  Req extends Request = Request,
  Res extends Response = Response
>(
  router: IRouter,
  operation: BridgeOperation<T>,
  schemaComponents: SchemaComponents,
  executor: GraphQLExecutor<Req, Res>,
  config?: CreateMiddlewareConfig<Req, Res>
) => {
  const route = pathTemplateToExpressRoute(operation.path)

  // Resolving `$ref`, at least OpenAPIRequestValidator cannot handle them properly
  const parameters_ = _.cloneDeep(operation.openAPIOperation.parameters || [])
  const requestBody_ = _.cloneDeep(operation.openAPIOperation.requestBody) as
    | RequestBodyObject
    | undefined

  resolveSchemaComponents(parameters_, schemaComponents)
  resolveSchemaComponents(requestBody_, schemaComponents)

  const requestCoercer = new OpenAPIRequestCoercer.default({
    parameters: parameters_,
    requestBody: operation.requestBodyFormData === 'FORM_DATA' ? requestBody_ : undefined,
  })

  const requestValidator =
    typeof config?.validateRequest !== 'boolean' || config.validateRequest
      ? new OpenAPIRequestValidator.default({
          parameters: parameters_,
          requestBody: requestBody_,
        })
      : undefined

  const responseValidator = config?.validateResponse
    ? new OpenAPIResponseValidator.default({
        // Type in `openapi-response-validator` seems wrong
        responses: operation.openAPIOperation.responses as any,
        components: { schemas: schemaComponents },
      })
    : undefined

  const graphqlDocument_ = print(operation.graphqlDocument)

  router[operation.httpMethod](
    route,
    promiseToHandler(async (req, res) => {
      const allRequestBodyVariables = Object.keys(operation.requestBodyVariableMap)

      if (operation.requestBodyFormData === 'JSON') {
        await jsonBodyParserPromise(req, res)
      }

      if (operation.requestBodyFormData === 'FORM_DATA') {
        await formBodyParserPromise(req, res)
      }

      const supportedContentTypes = Object.keys(
        (operation.openAPIOperation.requestBody as OpenAPIV3.RequestBodyObject)?.content || {}
      )
      if (
        ['POST', 'PUT', 'PATCH'].includes(req.method) &&
        supportedContentTypes.length > 0 &&
        !supportedContentTypes.includes(req.headers['content-type'] || '')
      ) {
        res.status(415).json({
          errors: [
            {
              message: `Only "${supportedContentTypes.join(', ')}" supported`,
            },
          ],
        })
        return
      }
      const req_ = {
        ...req,
        cookies: { ...req.cookies },
        headers: { ...req.headers },
        params: { ...req.params },
        query: { ...req.query },
        body: req.body,
      }

      // TODO: Cookies not supported
      requestCoercer.coerce(req_)

      const validationResult = requestValidator ? requestValidator.validateRequest(req_) : null

      if (validationResult) {
        res.status(400).json({ errors: validationResult.errors })
        return
      }

      let variables: Record<string, unknown> = {}
      for (const parameter of operation.openAPIOperation
        .parameters as OpenAPIV3.ParameterObject[]) {
        const variableName = operation.variableMap[parameter.name] || parameter.name
        if (parameter.in === 'path') {
          variables[variableName] = req_.params[parameter.name]
        }
        if (parameter.in === 'query') {
          variables[variableName] = req_.query[parameter.name]
        }
        if (parameter.in === 'header') {
          variables[variableName] = req_.headers[parameter.name]
        }
      }
      if (operation.requestBodyFormData === 'JSON') {
        for (const requestBodyVariable of allRequestBodyVariables) {
          variables[requestBodyVariable] =
            req_.body[operation.requestBodyVariableMap[requestBodyVariable]] ||
            (operation.requestBodyIsSingleInput ? req_.body : null)
        }
      }

      if (operation.requestBodyFormData === 'FORM_DATA') {
        for (const requestBodyVariable of allRequestBodyVariables) {
          variables[requestBodyVariable] =
            req_.body[operation.requestBodyVariableMap[requestBodyVariable]] || null
        }
      }

      if (operation.requestBodyFormData === 'MULTIPART_FORM_DATA') {
        variables = {
          ...variables,
          ...transformRequest(
            req,
            graphqlDocument_,
            allRequestBodyVariables,
            operation.requestBodyVariableMap
          ),
        }
      }

      const request = {
        document: graphqlDocument_,
        variables,
        request: req as Req,
        response: res as Res,
      }
      const result = await executor(request)

      if (config?.responseTransformer) {
        const response_ = await config.responseTransformer({
          result,
          request,
          openAPISchema: {
            operation: operation.openAPIOperation,
            method: operation.httpMethod,
            path: operation.path,
          },
        })
        if (response_) {
          const { statusCode, contentType, data } = response_
          if (contentType) {
            res.contentType(contentType)
          }
          res.status(statusCode).send(data).end()
          return
        }
      }

      // There are 2 cases of errors handled here;

      // 1. If responseValidation is enabled, it will throw in both cases
      // - If there is a response validation problem
      // - If there are any graphql errors
      if (responseValidator) {
        const responseValidationErrors = responseValidator.validateResponse(200, result.data)
        if (responseValidationErrors || result.errors) {
          throw new InvalidResponseError(
            responseValidationErrors?.message || 'GraphQL Error',
            responseValidationErrors?.errors || [],
            result.errors
          )
        }
      }
      // 2. If no response validation is enabled, we will return (possibly) both the errors and a (partial) result
      if (result.errors) {
        res.status(500).json(result).end()
        return
      }
      res.status(200).json(result.data).end()
    })
  )
}

const createExpressMiddleware = <
  T extends CustomOperationProps = CustomOperationProps,
  ThisRequest extends Request = Request,
  ThisResponse extends Response = Response
>(
  operations: BridgeOperations<T>,
  executor: GraphQLExecutor<ThisRequest, ThisResponse>,
  config?: CreateMiddlewareConfig<ThisRequest, ThisResponse>
) => {
  // TODO: Avoid depending on express directly?
  const router = express.Router()

  for (const operation of operations.operations) {
    addOperation(router, operation, operations.schemaComponents, executor, config)
  }

  return router
}

export const createOpenAPIGraphQLBridge = <
  ThisRequest extends Request = Request,
  ThisResponse extends Response = Response,
  T extends CustomOperationProps = CustomOperationProps
>(
  config: CreateOpenAPIGraphQLBridgeConfig
) => {
  const { graphqlSchema, graphqlDocument, customScalars } = config

  const graphqlSchema_ =
    typeof graphqlSchema === 'string' ? buildSchema(graphqlSchema) : graphqlSchema

  const graphqlDocument_ =
    typeof graphqlDocument === 'string' ? parse(graphqlDocument) : graphqlDocument

  const operations = getBridgeOperations<T>(graphqlSchema_, graphqlDocument_, customScalars)

  return {
    getExpressMiddleware: (
      executor: GraphQLExecutor<ThisRequest, ThisResponse>,
      config?: CreateMiddlewareConfig
    ) => createExpressMiddleware(operations, executor, config),
    getOpenAPISchema: (config: CreateOpenAPISchemaConfig<T>): OpenAPIV3.Document<T> =>
      createOpenAPISchemaWithValidate<T>(operations, config),
  }
}
