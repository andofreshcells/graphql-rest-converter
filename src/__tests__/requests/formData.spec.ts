import express from 'express'
import { buildASTSchema } from 'graphql'
import { gql } from 'graphql-tag'
import { addMocksToSchema } from '@graphql-tools/mock'
import { makeExecutableSchema } from '@graphql-tools/schema'
import { createOpenAPIGraphQLBridge } from '../../express.js'
import request from 'supertest'
import { createSchemaExecutor } from '../../graphQLExecutor.js'
import { URLSearchParams } from 'node:url'

const app = express()

const schema = buildASTSchema(gql`
  type Mutation {
    someMutation(id: Int!, someParameter: String!): String!
  }

  type Query {
    myQuery: Boolean
  }

  schema {
    mutation: Mutation
    query: Query
  }
`)

const gqlSchema = addMocksToSchema({
  schema: makeExecutableSchema({
    typeDefs: schema,
  }),
})
const bridge = createOpenAPIGraphQLBridge({
  graphqlSchema: gqlSchema,
  graphqlDocument: gql`
    mutation fetch($id: Int!, $someParameter: String! @OABody(contentType: FORM_DATA))
    @OAOperation(path: "/fetch") {
      someMutation(id: $id, someParameter: $someParameter)
    }
  `,
})

app.use(bridge.getExpressMiddleware(createSchemaExecutor(gqlSchema)))

describe('FormData', () => {
  it('should create the correct schema', () => {
    expect(
      bridge.getOpenAPISchema({
        baseSchema: {
          openapi: '3.0.3',
          info: {
            title: 'The API',
            version: '1.0.0',
            description: 'Description',
          },
        },
      })
    ).toMatchSnapshot()
  })
  it('should only support form-data requests', async () => {
    const result = await request(app).post('/fetch?id=5').expect(415)
    expect(result.body).toMatchSnapshot()
  })
  it('should validate form data requests', async () => {
    const result = await request(app).post('/fetch?id=5').type('form').expect(400)
    expect(result.body).toMatchSnapshot()
  })
  it('should POST `/fetch` successful', async () => {
    const form = new FormData()
    form.append('someParameter', 'nice')
    const result = await request(app)
      .post('/fetch?id=5')
      .type('form')
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      .send(new URLSearchParams(form).toString())
    expect(result.body).toMatchSnapshot()
  })
})
