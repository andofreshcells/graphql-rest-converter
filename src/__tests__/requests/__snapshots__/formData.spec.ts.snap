// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormData should POST \`/fetch\` successful 1`] = `
{
  "someMutation": "Hello World",
}
`;

exports[`FormData should create the correct schema 1`] = `
{
  "components": {
    "schemas": {},
  },
  "info": {
    "description": "Description",
    "title": "The API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": {
    "/fetch": {
      "post": {
        "operationId": "fetch",
        "parameters": [
          {
            "in": "query",
            "name": "id",
            "required": true,
            "schema": {
              "format": "int32",
              "type": "integer",
            },
            "x-graphql-variable-name": "id",
          },
        ],
        "requestBody": {
          "content": {
            "application/x-www-form-urlencoded": {
              "schema": {
                "properties": {
                  "someParameter": {
                    "type": "string",
                    "x-graphql-variable-name": "someParameter",
                  },
                },
                "required": [
                  "someParameter",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "someMutation": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "someMutation",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "errorCode": {
                            "type": "string",
                          },
                          "location": {
                            "type": "string",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "415": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "message": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Unsupported Media Type",
          },
          "500": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "data": {
                      "description": "Branch of data that does not contain errors",
                      "nullable": true,
                      "properties": {
                        "someMutation": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "someMutation",
                      ],
                      "type": "object",
                    },
                    "errors": {
                      "items": {
                        "properties": {
                          "locations": {
                            "items": {
                              "properties": {
                                "column": {
                                  "type": "integer",
                                },
                                "line": {
                                  "type": "integer",
                                },
                              },
                              "type": "object",
                            },
                            "type": "array",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "items": {
                              "type": "string",
                            },
                            "type": "array",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal Server Error",
          },
        },
        "x-graphql-operation": "mutation fetch($id: Int!, $someParameter: String!) {
  someMutation(id: $id, someParameter: $someParameter)
}",
      },
    },
  },
}
`;

exports[`FormData should only support form-data requests 1`] = `
{
  "errors": [
    {
      "message": "Only "application/x-www-form-urlencoded" supported",
    },
  ],
}
`;

exports[`FormData should validate form data requests 1`] = `
{
  "errors": [
    {
      "errorCode": "required.openapi.requestValidation",
      "location": "body",
      "message": "must have required property 'someParameter'",
      "path": "someParameter",
    },
  ],
}
`;
